### ISSUE TITLE
Istio ingress still forwarding request to a pod that got marked terminated 9mins ago

### ISSUE THREAD

**Issue by ra-grover** [istio/istio#44985]
### Bug Description

Hi Team,
We are running Istio (1.17.1-distroless) in our EKS cluster, we found out that during the event when a replicaset scaled down rapidly, from 350 Pods to 180 in around 10 mins, we faced a lot of 503s. 
This we beleive is due to the fact that the istio ingressgateway still makes a request to the terminated pod's proxy, even after 3 minutes of the Endpoints update. 
The Pod count in the EKS cluster is 3K and Nodes being ~250. There are about 200 services too. 

Here is a chronology of the events:
1. The pod(*-rbzcp) got marked for deletion by the replicaset controller at `2023-05-17T02:39:02.754863Z`
2. The IP of the pod got updated in the endpoint `2023-05-17T02:39:02.891734Z`
3. The pod got deleted from the node's kubelet at `2023-05-17T02:42:08.759850Z`

The istio proxy logs on the same pod:
```2023-05-17T02:42:07.910650Z	info	Graceful termination period complete, terminating remaining proxies.
2023-05-17T02:42:07.910757Z	warn	Aborted proxy instance
2023-05-17T02:42:07.910766Z	warn	Aborting proxy
2023-05-17T02:42:07.910953Z	info	Envoy aborted normally
2023-05-17T02:42:07.910966Z	info	Agent has successfully terminated
2023-05-17T02:42:07.911016Z	error	error serving tap http server: http: Server closed
2023-05-17T02:42:07.911374Z	info	ads	ADS: "@" weaver-loom-prod-us-west-2-bex-template-7dbdcdc7f6-rbzcp.experience-api-1 terminated
2023-05-17T02:42:07.911388Z	info	ads	ADS: "@" weaver-loom-prod-us-west-2-bex-template-7dbdcdc7f6-rbzcp.experience-api-2 terminated
2023-05-17T02:42:07.911402Z	info	sds	SDS server for workload certificates started, listening on "./var/run/secrets/workload-spiffe-uds/socket"
```
There is no log from the pod in question(*-rbzcp), after this. 
The image shows that after marked terminated, the ingress continued to forward requests to the pod, getting 503s.
<img width="959" alt="Screenshot 2023-05-18 at 17 16 16" src="https://github.com/istio/istio/assets/93419161/74021553-c2fa-4c90-9442-e26b5e840259"> 

At this time, the pod count of this deployment reduced as shown below (HPA scale-in), from 330 to 180 over 15 mins
<img width="289" alt="Screenshot 2023-05-18 at 17 45 28" src="https://github.com/istio/istio/assets/93419161/e3e5601d-b5c5-4322-af26-1ebaa399284e">

We are running 50 replicas of Istiod but they seem to throttled at that time (please not that throttling is there at other occurrences too, but we are not seeing 503/outages)
<img width="701" alt="Screenshot 2023-05-18 at 17 50 20" src="https://github.com/istio/istio/assets/93419161/94064797-7f51-404e-8ede-27a2ba410e17">

### Version

```Text
istioctl version:
client version: 1.15.3
control plane version: 1.17.1
data plane version: 1.17.1 (59 proxies)

 kubectl version --short
Client Version: v1.23.6
Server Version: v1.23.16-eks-48e63af

helm version --short
v3.6.3+gd506314
```

### Additional Information

Istiod logs for that pod:
```
2023-05-17T02:42:07.911631Z	info	ads	ADS: "10.64.5.57:36606" <REDACTED>-rbzcp.<REDACTED>-27184 terminated
2023-05-17T02:32:53.743852Z	info	ads	RDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:21 size:1.6MB cached:10/21
2023-05-17T02:32:52.883771Z	info	ads	NDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:1 size:27.2kB
2023-05-17T02:32:51.939697Z	info	ads	LDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:63 size:291.3kB
2023-05-17T02:32:51.929958Z	info	ads	EDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:587 size:1.1MB empty:0 cached:586/587
2023-05-17T02:32:50.755431Z	info	ads	ADS: "10.64.5.57:48574" <REDACTED>-rbzcp.<REDACTED>-27024 terminated
2023-05-17T02:32:51.001296Z	info	ads	CDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:654 size:837.2kB cached:644/650
2023-05-17T02:32:50.995933Z	info	ads	ADS: new connection for node:<REDACTED>-rbzcp.<REDACTED>-27184
2023-05-17T02:23:28.027324Z	info	ads	NDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:1 size:27.2kB
2023-05-17T02:23:28.023420Z	info	ads	RDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:21 size:1.6MB cached:12/21
2023-05-17T02:23:28.001766Z	info	ads	LDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:63 size:291.3kB
2023-05-17T02:23:27.996641Z	info	ads	EDS: PUSH INC for node:<REDACTED>-rbzcp.<REDACTED> resources:2 size:155.6kB empty:0 cached:0/2
2023-05-17T02:23:27.167235Z	info	ads	CDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:654 size:837.2kB cached:650/650
2023-05-17T02:18:49.324106Z	info	ads	NDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:1 size:27.2kB
2023-05-17T02:18:49.320813Z	info	ads	RDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:21 size:1.6MB cached:12/21
2023-05-17T02:18:49.297607Z	info	ads	LDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:63 size:291.3kB
2023-05-17T02:18:49.291275Z	info	ads	EDS: PUSH INC for node:<REDACTED>-rbzcp.<REDACTED> resources:1 size:143.5kB empty:0 cached:0/1
2023-05-17T02:18:48.444070Z	info	ads	CDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:654 size:837.2kB cached:645/650
2023-05-17T02:16:20.119399Z	info	ads	NDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:1 size:27.2kB
2023-05-17T02:16:20.115814Z	info	ads	RDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:21 size:1.6MB cached:12/21
2023-05-17T02:16:20.091425Z	info	ads	LDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:63 size:291.3kB
2023-05-17T02:16:20.084363Z	info	ads	EDS: PUSH INC for node:<REDACTED>-rbzcp.<REDACTED> resources:1 size:190.7kB empty:0 cached:0/1
2023-05-17T02:16:19.217966Z	info	ads	CDS: PUSH for node:<REDACTED>-rbzcp.<REDACTED> resources:654 size:837.2kB cached:650/650
2023-05-17T02:02:31.076854Z	info	ads	ADS: "10.64.5.57:34598" <REDACTED>-rbzcp.<REDACTED>-27211 terminated
2023-05-17T02:02:31.207645Z	info	ads	RDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:21 size:1.6MB cached:12/21
2023-05-17T02:02:31.181192Z	info	ads	NDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:1 size:27.2kB
2023-05-17T02:02:31.180526Z	info	ads	LDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:63 size:291.3kB
2023-05-17T02:02:31.171874Z	info	ads	EDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:587 size:1.1MB empty:0 cached:587/587
2023-05-17T02:02:31.167626Z	info	ads	CDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:654 size:837.2kB cached:629/650
2023-05-17T02:02:31.162313Z	info	ads	ADS: new connection for node:<REDACTED>-rbzcp.<REDACTED>-27024
2023-05-17T01:31:21.527483Z	info	ads	RDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:21 size:1.6MB cached:12/21
2023-05-17T01:31:21.482855Z	info	ads	NDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:1 size:27.2kB
2023-05-17T01:31:21.482188Z	info	ads	LDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:63 size:291.3kB
2023-05-17T01:31:21.473726Z	info	ads	EDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:587 size:1.1MB empty:0 cached:587/587
2023-05-17T01:31:21.470405Z	info	ads	CDS: PUSH request for node:<REDACTED>-rbzcp.<REDACTED> resources:654 size:837.2kB cached:624/650
2023-05-17T01:31:21.465776Z	info	ads	ADS: new connection for node:<REDACTED>-rbzcp.<REDACTED>-27211
2023-05-17T01:31:21.368391Z	info	ads	ADS: "10.64.5.57:33552" <REDACTED>-rbzcp.<REDACTED>-26893 terminated


```

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [X] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

### Is this the right place to submit this?

- [X] This is not a security vulnerability
- [X] This is not a question about how to use Istio

**Comment by howardjohn** [comment]
what we should see logged in istiod is a `debounce ... ServiceEntry` when it gets endpoints events. do you see that?

Our best clue will be istiod logs, the truncated /redaction will limit that

9 minutes seems egregiously long, not the result of a slow update from k8s or throttling unless things are extremely loaded. Do we have any API server metrics that may indicate load on that end?

Basically somewhere from API server to istiod to proxy the endpoints update is lost, so trying to establish where.

**Comment by paulhamby** [comment]
>unless things are extremely loaded

They are very loaded and that's part of the concern; the istiod instances seem busier than I would expect for longer than I'd expect.

Many minutes prior to this one case that @ra-grover has pointed out, there was an application deployment that caused about 700 application pods to get replaced with a newer version. There was a burst of events from K8s that were received by istio (mainly of type:pod. event:update).
![Screenshot 2023-05-19 at 3 26 06 PM](https://github.com/istio/istio/assets/4423579/7223321f-3f0c-4a4e-9e72-7865fd7b4f8a)

Which then resulted in a lot of istio pushes, but much more than I would have expected and for longer:
![Screenshot 2023-05-19 at 3 24 03 PM](https://github.com/istio/istio/assets/4423579/eac0475a-8046-42ad-b25e-655bf0a4ca83)

And while these pushes are occurring the istiod CPU usage was _very_ high:
![Screenshot 2023-05-19 at 3 27 42 PM](https://github.com/istio/istio/assets/4423579/522aaead-9613-4eeb-b1d4-ab9c0d7f5048)

And the Convergence time was increased:
![Screenshot 2023-05-19 at 3 35 24 PM](https://github.com/istio/istio/assets/4423579/8780fddc-ca75-495d-bb50-dd85e7575f41)


I would expect that the high number of pod deletes and creates results in a large amount of work to be done, but it seems that the number of pushes is potentially disproportionate to the number of K8s events that precede it.

All times in screenshots above are in UTC.

**Comment by paulhamby** [comment]
It seems to me that, at least in our environment, istio is pushing more configuration changes than I would expect given the amount of change in the system. I gathered up istio logs and EKS audit events from the relevant time frame above and did some analysis.

In the hour from 2023-05-17T02:00 to 2023-05-17T03:00 UTC there were 16,838 events that mutated (create, update, delete) endpoints, endpointslices, pods, or services, which seems like a pretty high rate of change. 11,145 were for a fairly large service that did a deploy during that time. During this hour there were 276,486 total istio pushes.

Here is a graph of those pushes and events on a per minute basis.
<img width="1312" alt="events_vs_pushes" src="https://github.com/istio/istio/assets/4423579/ba11d3db-acd2-41d1-881c-690e3dcca32e">

I suppose what's most concerning is the exponential growth of pushes after the spike of events around the 35 minute mark. Although the overall ratio of pushes to config changes is also curious.

Any clues as to what might cause this? Or an area we should focus our investigation on?

The majority of the events are updates to endpointslices. Unfortunately the EKS audit policy logs those at metadata level, so we don't get request/response data in the logs.

Our hypothesis is that istio is lagging behind in processing and pushing configurations, resulting in stale configurations and requests being sent to dead pods and 503s for users. While delays in the seconds range would be understandable, we're periodically seeing requests sent to dead pods that have been dead for as long as 10-20 minutes. We've scaled istio up and out, but our thinking is that we've either hit a bug in istio or there's something particular about our environment that's causing this behavior.

**Comment by howardjohn** [comment]
Just double checking, is the push metric number of pushes _to a proxy_? Since in the naive case there would be events*proxies pushes measuring that way (we optimize it a bit, so it should be less). If its pilot_push_triggers its not quite fair to treat it as 1 "push" as it may be multiple pushes consolidated into one.

16,838 events is pretty wild, that is 4 updates per second sustained for an hour (and its really spikier, as noted). Even ignoring k8s that seems like quite a bit. I guess Pod updates a lot and Istio filters those out cheaply

Would be more clear to look at:

pilot_pushcontext_init_seconds_count: full pushes only (more expensive)
pilot_debounce_time_count: all pushes (includes full pushes, but also endpoint-only updates which are cheap)
pilot_proxy_queue_time_count: all pushes per proxy


I would expect you to see high number of endpoint only updates,which should be quite cheap

**Comment by paulhamby** [comment]
The pushes per minute value was derived from logs by `for i in {00..59}; do; grep "Pushing" combined_istio.log | grep ^2023-05-17T02:$i | wc -l; done `. Since I was working with a .csv export of EKS audit events, I found it would be beneficial to also work with the istio logs during that time period, rather than trying to match up with metrics in our metrics platform.

Unfortunately we aren't currently recording the metrics you mention.

**Comment by howardjohn** [comment]
Pushing logs will be inflated by number of istiods, so it is not too weird for it to be larger than events. Each event causes each istiod to push to all connected proxies (note: proxy connects to 1 istiod at a time).

Grepping for this log `Push debounce stable[3] 1 for reason unknown: 2.34µs since last change, 2.3µs since last push, full=true` will show the full vs not full part

**Comment by paulhamby** [comment]
Ok, hopefully this is more accurate:

Full Pushes:
`for i in {00..59}; do; grep "Push debounce" istio_sorted.txt | grep "full=true" | grep ^2023-05-17T02:$i | wc -l; done`
Partial Pushes:
`for i in {00..59}; do; grep "Push debounce" istio_sorted.txt | egrep -v "full=true" | grep ^2023-05-17T02:$i | wc -l; done`

<img width="1100" alt="Screenshot 2023-05-24 at 4 46 32 PM" src="https://github.com/istio/istio/assets/4423579/d46e4fda-6194-4007-ac9e-f66575ded9ac">

There were 50 full pushes at minute 17 and 19, 150 in total, and here are the minutes where there was a full push (hard to see in the graph).

```
16      50
18      50
21       1
22       5
23      18
24       7
25       7
26       6
27       3
28       2
29       1
```

**Comment by paulhamby** [comment]
Thanks for the tips. We've been looking at the sidecar for a mid to long term solution, will just take us a bit to implement. We've also just scaled up some more, so glad to know we're thinking about it the same. PILOT_DEBOUNCE_AFTER is one of the params we've tried adjusting, but haven't yet had it help, we're back at the default now but will revisit.

As far as the number of pushes go, even though they are partial pushes, doesn't the number of pushes still seem too great? The ratio of pushes to events just seems off.

Also, that exponential increase in pushes at that 35 minute mark in this example is still concerning.

**Comment by howardjohn** [comment]
If you saw no merging/debouncing, I would expect Events*Istiod instances pushes, so it would be more than events. At worst you look to have 7x as many pushes as events, and you have 50 Istiods, so that doesn't seem implausible

**Comment by ra-grover** [comment]
Hi @howardjohn 
We spent some time digging into the issue and found out that the Controller queue (instance of [this](https://github.com/istio/istio/blob/master/pkg/queue/instance.go#L64)) grows to very high numbers. 3000 items at times, we added some metrics for the length of the queue [here in our fork](https://github.com/ra-grover/istio/blob/ef0ffae0fed3776040266f16bc6cd66561e46dbb/pkg/queue/instance.go#L140). 

Since this queue is only dequeued one at a time by a single goroutine, it took crazy amount of time to get cleared (5-10 mins) during moments of churn (Rolling deployments of 500 Replicas).

We then added a metric here to observe the time it takes to call a `task()` , which is dequeued from the queue and called, and noted that the endpoint-updates took relatively large time during moments of churn. It went as high as to 0.9 seconds to call the dequeued callback. We are observing that [here](https://github.com/ra-grover/istio/blob/ef0ffae0fed3776040266f16bc6cd66561e46dbb/pkg/queue/instance.go#L140) in our fork.

Please note that rather than enqueuing callbacks to queue, we are pushing an enriched version of that [here](https://github.com/ra-grover/istio/blob/7a9824a9ba1715acb5ec0b62fde672cfce76fc44/pilot/pkg/serviceregistry/kube/controller/controller.go#L692) and also other places this queue is used, so that we are able to count the number of objects in the queue by the type, like `endpoint-update`, `pod-create`, `pod-update` etc and also note their enqueue times. 

Next we found that the `endpointslices-update` took the highest time (after dequeuing, while executing those callbacks)
![Screenshot 2023-05-25 at 11 24 17](https://github.com/istio/istio/assets/93419161/94a7c398-7fba-4f01-98ca-cc037fb8c78d)



We then wrapped duration logs to gather time durations spent `processEndpointEvent` and it turns out that at times of churn the `s.edsCacheUpdate`  took a significant amount of time. 

Here are some logs, please relate them to our fork [here](https://github.com/ra-grover/istio/blob/08443dba42579f7272700ac7ea72d2d753e68e7b/pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go#L71) , [here](https://github.com/ra-grover/istio/blob/08443dba42579f7272700ac7ea72d2d753e68e7b/pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go#L127) and [here](https://github.com/ra-grover/istio/blob/08443dba42579f7272700ac7ea72d2d753e68e7b/pilot/pkg/xds/eds.go#L82) .
```
2023-05-25T18:27:38.892622Z	info	Dequeuing task EndpointSlice-update , process time 983314289 (microsecs), queue length: 376
2023-05-25T18:27:38.892645Z	info	kube	Handle EDS endpoint stark-grpc-rcp-showcase update in namespace stark
2023-05-25T18:27:38.992571Z	info	ads	Push debounce stable[1538] 1 for config ServiceEntry/stark/stark-grpc-rcp-showcase.stark.svc.cluster.local: 100.254352ms since last change, 100.254122ms since last push, full=false
2023-05-25T18:27:38.992656Z	info	ads	XDS: Incremental Pushing:2023-05-25T18:08:10Z/2 ConnectedEndpoints:173 Version:2023-05-25T18:08:10Z/2
2023-05-25T18:27:39.801443Z	info	ads	The s.ConfigUpdate took 0.000015 secs
2023-05-25T18:27:39.801547Z	info	Dequeued task EndpointSlice-update , took 0.908907 (secs)
2023-05-25T18:27:39.801567Z	info	kube	The XDSUpdater.EDSUpdate took 0.906140 secs
2023-05-25T18:27:39.801613Z	info	kube	serviceLister took 0.000072 (secs)
2023-05-25T18:27:39.801568Z	info	kube	updateEDs took 0.908905 (secs)
```

The
```
The XDSUpdater.EDSUpdate took 0.906140 secs
```
contains `s.edsCacheUpdate` to calculate the pushType. Its ironical that to just infer the type of push it takes so much time. Because the other part `s.ConfigUpdate` takes only ` 0.000015 secs` in `XDSUpdater.EDSUpdate ` as mentioned in logs above.

As you can see this takes almost a second to complete and the controller queue reaches to a significant number. We dont have this data during our incidents, we have tried to create a churn in our test environments trying to find the contentions. 

Few questions:
1. Can we have a metric officially for the length of the queue ?
2. Why is the controller queue being dequeued one at a time ? Is it intentional that there are no multiple go-routines handling dequeuing and calling the callbacks. 
3. Is there a possible optimization possible in `s.edsCacheUpdate` ?

We are up for a contribution if you want to expose both these metrics (if maintainers want it). 

We are also in process of gathering this data, both these metrics (length of queue and time taken to execute a task), in our production environments and see if this was the real cause for 9 mins delay that we specified in the issue.

**Comment by howardjohn** [comment]
Will respond to the rest soon (in a meeting), but in the meantime were you
able to get a CPU profile at all?
https://github.com/istio/istio/wiki/Analyzing-Istio-Performance#control-plane
I think it will help explain somethings more clearly

**Comment by tehlers320** [comment]
@howardjohn same company, this is from a generically loaded lab though. 
[pprof.pilot-discovery.samples.cpu.001.pb.gz](https://github.com/istio/istio/files/11569536/pprof.pilot-discovery.samples.cpu.001.pb.gz)

**Comment by howardjohn** [comment]
One question before we get into it: have you used other Istio versions before and done similar things? if so, did you see the same issue? Trying to understand if its a regression or has always been any issue.

---

So first want to acknowledge our current controller is not ideal - definitely room for improvements we are looking into.

The single thread queue is intentional. We rely on it to preserve correct ordering between events. See https://github.com/istio/istio/issues/45114 for a recent (minor) bug found due to _not_ using the queue correctly.

There was a PR to add queue metrics (https://github.com/istio/istio/pull/40398) but it didn't move forward. Possibly could be revived.

We had a previous issue like this long ago; it ended up being that in the queue we were making synchronous calls to the API server. This is rate limitted (and has latency in general), so it result in basically this exact issue - queue backed up.

Since then, we have removed (as far as I know) all of these calls. All queue processing should be CPU bound. With your numbers, that means we are processing 10 events per second, so 100ms per event. In your logs you showed some even worse. That seems awfully slow; CPU profile may help explain why. Seems like either we are hitting some bad case (some n^2 algorithm?) or we are so severely throttled from the XDS pushing that we fall behind here a lot.

---

Not sure if all the time is spent on EDSUpdate or that is just one example, but I will focus on that. Really, profiles would help a ton - the logging is excellent but profiles are the best source of info here. Getting a mutex one may help as well.

> Its ironical that to just infer the type of push it takes so much time

FWIW it is doing more than just checking, it also updates the indexes that are used later. 

First, can you check if you have PILOT_SEND_UNHEALTHY_ENDPOINTS=true set? It defaults to false. I doubt this is it, just to double check though.

The next most obvious thing is lock contention. Lots of locking involved in this flow. One thing we can do to reduce it is to try PILOT_ENABLE_XDS_CACHE=false if you are up for it. Not this will probably _dramatically_ increase CPU usage, so may be a bad idea. But there is a fair amount of locking here, so turning it off could give some info.

The locking bit is related to https://github.com/istio/istio/issues/40744 perhaps. There have been some improvements (or changes, at least - hopefully they help :slightly_smiling_face: ) to the locking around the cachewhich may or may not help. So trying 1.18.0-rc could be useful perhaps.

I will try to find some time to reproduce (likely using https://github.com/howardjohn/pilot-load/), but no promises it will be soon unfortunately

**Comment by tehlers320** [comment]
We are using https://github.com/kubernetes/perf-tests/tree/master/clusterloader2 to generate artifical load so its not 100% what we have in real environments. 

```
kubectl get virtualservice -A |wc -l
    1636
```

This is more realistic of a cluster
```
kubectl get virtualservice -A |wc -l
     295
```

That bit of code you linked i modified it to have values as long as ours and it didnt make much of an impact on the performance.

```
goos: darwin
goarch: arm64
BenchmarkNameMatch
BenchmarkNameMatch-10            	25333431	        47.10 ns/op	       0 B/op	       0 allocs/op
BenchmarkNameMatch_Ourworkload
BenchmarkNameMatch_Ourworkload-10    	24597325	        49.05 ns/op	       0 B/op	       0 allocs/op
PASS
```

**Comment by ra-grover** [comment]
Here is one more profile take when the queue length is 4900 in one of the other clusters. 

[pprof.pilot-discovery.samples.cpu.012.pb.gz](https://github.com/istio/istio/files/11569702/pprof.pilot-discovery.samples.cpu.012.pb.gz)

**Comment by ra-grover** [comment]
>  Will drop CPU usage quite a bit, unless you have your own EnvoyFilters you use. Also in 1.18

We are using our own filters.

**Comment by howardjohn** [comment]
In theory based on 50ns per op, you are calling NameMatch 200million times in 30s. Thats either a surprise N^2 somewhere or its actually slower than 50ns in practice. Or the pprof is just inaccurate for some reason.

The new profile is entirely different. 100% endpoint pushes (other had CDS/RDS/LDS as well), so must be only endpoints changing at the time.

About 50% time on marshalling (expected), 50% time on https://github.com/istio/istio/blob/956dd0cf37c50643bcfc197fc55bcc3ea38113a8/pilot/pkg/xds/endpoint_builder.go#L322 (seems higher than I would expect). I think that can be optimized pretty easily, will put up a PR. I don't think its the cause of the issue though

**Comment by ra-grover** [comment]
> On the same cluster, from which the above profile is shared,
> kube	The XDSUpdater.EDSUpdate took 2.107840 secs

@howardjohn I attached a debugger remotely to the same cluster. I definitely observed a lag when I put a breakpoint before and after. 
https://github.com/ra-grover/istio/blob/08443dba42579f7272700ac7ea72d2d753e68e7b/pilot/pkg/xds/eds.go#L127

I can try wrapping durations here.

**Comment by ra-grover** [comment]
Yes, it is that line only. 700 milliseconds to just acquire the lock! 
```ads	Time taken to ep.Lock(): 0.716286```, The queue length was just 200 not something crazy like 4000. Trying to create that churn.

**Comment by ra-grover** [comment]
Also just noticed one more behaviour, the queue takes massive time to get cleared in order of minutes. But when it comes to clearing (not sure what gets triggered ), it clears very very fast! From 1300 to 0, it came down in less than 1.5 seconds.

**Comment by howardjohn** [comment]
any idea if there is a single task that is taking a super long time and blocking it? otherwise it's not getting scheduled I would imagine.

How many cores does the node istiod run on have? that will impact GOMAXPROCS.

I don't want to keep asking you to try random things, but if you are in the experimenting mood it may be interesting to run 1 istiod instance with 0 proxies connected (can just not put it in the service or something) and see if it experiences the same thing . I suspect it would not

**Comment by ra-grover** [comment]
> but if you are in the experimenting mood it may be interesting to run 1 istiod instance with 0 proxies connected (can just not put it in the service or something) and see if it experiences the same thing . I suspect it would not

I ran multiple times from my local (connecting to EKS) and since it has 0 connected endpoints, I was not able to get the queue to grow.  But I can give it a try in the actual cluster with 0 proxies connected tomorrow. 

Node varies, on test where I reproed this contention in acquiring the lock, the machine is `c6a.12xlarge` but I can see multiple istiods scheduled on the same one. 
While on prod where the incident occurred, it is `m6g.4xlarge` but only two istiod pods are scheduled.

**Comment by paulhamby** [comment]
I swore I saw that automaxprocs was in use, but I guess I mis-read a github issue. Would you recommend setting to the value of requests or limits?

**Comment by howardjohn** [comment]
I haven't done any evaluation of the impacts on Istio so I have no real recommendations, but its worth experimenting with setting it to the limit value I think

**Comment by ra-grover** [comment]
Here is one Mutex profile at the time of churn. `1m5s` of mutex time. Not sure if this helps you in some way. 
<img width="1066" alt="Screenshot 2023-05-26 at 11 57 04" src="https://github.com/istio/istio/assets/93419161/2204d291-010d-4bb4-b0ca-f329f9b5f3cb">

**Comment by ra-grover** [comment]
The corresponding CPU profile for the above shared Mutex profile(1min5sec) is below:

<img width="1046" alt="Screenshot 2023-05-26 at 12 29 50" src="https://github.com/istio/istio/assets/93419161/3bf274ab-066f-403f-b4b1-3ce9d637b2cd">

@howardjohn So looks like Instance.String is being the issue, [here](https://github.com/ra-grover/istio/blob/68fa9070813a5724516b98b252b7e7c3e3eddc09/pkg/config/labels/instance.go#L139) ?

**Comment by ra-grover** [comment]
@howardjohn Apologies for my ignorance on the codebase, but whats the idea behind combining all those labels to a string and using it as a key [here](https://github.com/istio/istio/blob/master/pilot/pkg/xds/mtls_checker.go#L100) ?

If thats being used to uniquely identify a workload, can a combination of Endpoint Address and Pod name (from labels), be be used for the same purpose ?

**Comment by howardjohn** [comment]
I think the idea was to reduce computation because most endpoints in the same service will have the same labels. Its plausible that costs more than it saves, though.

I put together another patch that reduces locking here a lot. It drops from a Lock to a RLock, and does all computation outside the lock.

https://github.com/istio/istio/compare/release-1.17...howardjohn:istio:xds/drop-locking?expand=1. Note there is 2 commits, I would apply both, the first was just a slightly simpler attempt.

Note both of them result in higher CPU usage, but should have less lock contention. new=1 commit, new2=both commits

```
name \ time/op               /tmp/old     /tmp/new      /tmp/new2
EndpointGeneration/1000/1-8  2.41ms ± 0%  31.09ms ± 0%   2.85ms ± 0%

name \ alloc/op              /tmp/old     /tmp/new      /tmp/new2
EndpointGeneration/1000/1-8   695kB ± 0%   8287kB ± 0%   1870kB ± 0%

name \ allocs/op             /tmp/old     /tmp/new      /tmp/new2
EndpointGeneration/1000/1-8   12.1k ± 0%   208.1k ± 0%    31.1k ± 0%
```

**Comment by ra-grover** [comment]
Thanks for putting that out, we will try it. 
In our case, the labels were containing the pod specific nuances, like even the `pod-template-hash`, so I think it was definitely defeating the purpose of both the caching and on top of it the overhead of computing the labels.

**Comment by howardjohn** [comment]
Aside from larger changes one short term win may be to filter it to only
labels referenced by policies. Which is likely a small number

**Comment by ra-grover** [comment]
I replaced peerAuthnKey to be like `peerAuthnKey := ep.Address + ":" + ep.Labels["pod-template-hash"]`, which removed `ep.Labels.String()`, we are still seeing delays! 

Will try out your patch next week. 

> Aside from larger changes one short term win may be to filter it to only
> labels referenced by policies. Which is likely a small number

Totally agree with it, if I am understanding right, was the cache mainly intended to protect:
```c.peerAuthDisabledMTLS[peerAuthnKey] = factory.
NewPolicyApplier(c.push, ep.Namespace, ep.Labels).
GetMutualTLSModeForPort(ep.EndpointPort) == model.MTLSDisable
```

and its not being protected, because the labels in `ep` are pod specific. And may be as you said we need to filter out labels by service or some other criteria.

**Comment by paulhamby** [comment]
We were able to get a trace from our production environment that I think shows that we're on the right track addressing the contention. The 26 goroutines that account for 56% of execution time (trace was taken for 5 seconds while the istiod process was super busy burning cpu). Each goroutine was blocked on `*sync.RWMutex).Lock()` for 4.5-4.7 seconds of the 5 seconds.

![Screenshot 2023-05-26 at 6 39 04 PM](https://github.com/istio/istio/assets/4423579/a0f9ae31-dff8-4480-8afa-4f2a80a96208)
![Screenshot 2023-05-26 at 6 36 19 PM](https://github.com/istio/istio/assets/4423579/209d1edf-ace3-489c-9499-a9ea97c7e082)

**Comment by ra-grover** [comment]
Let me check that in some other cluster and also compare the controller queue processing durations with the other version.

**Comment by rahul656** [comment]
We have some interesting results to share with the `PILOT_PUSH_THROTTLE` config set to`5`.  Tests are done by restarting an app with 500 pods.

- We don't see the events piling up in the work queue, whereas, with the default `pilot_push_throttle` config, endpointSlice events depth reaches up to 1.5K. 
- Updates are sent across proxies almost within the time shown by `Debounce` + `Convergence time`, whereas with default config it took approx 25 to 30 mins. 
- Time taken to acquire `[ep.lock](https://github.com/istio/istio/issues/44985#issuecomment-1568735005)` reduced from Max 1.2 seconds to 52ms. 
- Pushes throughput and latency improve as well. 

I've attached a few other stats below comparing the two.

Pilot_push_throttle - 100
<img width="1740" alt="PILOT_PUSH_THROTTLE-100" src="https://github.com/istio/istio/assets/28567386/f10657f4-9bf7-4537-992d-e409d4014f2e">

Pilot_push_throttle - 5
<img width="1740" alt="PILOT_PUSH_THROTTLE-5" src="https://github.com/istio/istio/assets/28567386/fa4072f8-d260-4b63-ba75-757cbcc98d66">


 How is reducing the push_throttle improve the rate at which events are processed? Isn't the event processing supposed to be decoupled from the pushes?

**Comment by howardjohn** [comment]
The problem is the pushes and event processing both access the same lock. With PILOT_PUSH_THROTTLE=100 we have 100 things reading the lock and 1 writing (with my patch; with the release code all are getting write locks). With PILOT_PUSH_THROTTLE=5 we only have 5 things reading the lock so there is less contention

**Comment by rahul656** [comment]
Given the fact that we see considerable improvement after reducing the Pilot_push_throttle to `5` during the tests, was wondering if there is any other caveat that we need to be aware of before we apply the change in production clusters.

If higher concurrent pushes - 100 impacts the pilot's event processing rate, should not the default concurrency be lower?

**Comment by howardjohn** [comment]
I don't think there has been much extensive testing around tuning this variable (at least not that I was involved in or aware of, anyhow). It was set to 100 4 years ago by me, and IIRC I just picked an arbitrary number: https://github.com/istio/istio/commit/3a796c56239296b4442235540e364196f49a5cd3.

In general, I am a bit skeptical this variable is ideal for tuning. A "push" has 2 components - generating the response (CPU bound) and sending the response (IO/network bound), but is bound by the same thing. So too high a value likely _decreases_ efficiency of generating response (I would expect throttle ~= threads is ideal). however, it may mean that a few IO bound pushes could block an unexpected number of proxies. In the past we had a rare bugs where the push would take forever, for example;  with 100 spots open that is not so bad if we lose 5, if we have 5 then we deadlock everything. Of course, those bugs have been fixed -- but you can still have slow proxies that clog things up. 

So overall I think:
1. 5 is probably reasonable for your case, although a slightly higher number like 10 or 20 may be good enough and less divergence from general population
2. Its not acceptable the performance you are seeing; _something_ should change out of the box
3. There is no yet enough evidence that reducing PILOT_PUSH_THROTTLE to 5 by default is generally safe or the correct fix for (2)

One thing to note re: defaults is this is a somewhat niche case of quickly scaling down a very large service. We should not generalize the solution to that to the default without sufficient testing. Part of that testing could be you rolling it out in production and reporting back that it is great (more than great - there is nothing more valuable than real world data for performance info), but we will also want to do some testing on other workloads, as part of the projects own testing infra, etc.

**Comment by ra-grover** [comment]
> Let me check that in some other cluster and also compare the controller queue processing durations with the other version.

I did saw better results with your patch in our test environments, I monitored for the time taken to call the dequeued controller callbacks. There are few cases where with the patch it still takes 0.1 seconds but that count is considerably less as compared to when I am running without your patch.

**Comment by rahul656** [comment]
thanks @howardjohn. To add to what @ra-grover mentioned in his last comment and share the stats with the patch and  `push_throttle` set to 100 and 10. The patch has reduced the lock contentions significantly and now we do not see much difference in event processing with different throttle configs, however, we do see that `push_times` are slightly better keeping the throttle to around 10. 

<img width="1740" alt="Lock-contention-fix" src="https://github.com/istio/istio/assets/28567386/33e28e32-f0e2-49e2-9d46-7b2e1e9fcf23">

**Comment by hzxuzhonghu** [comment]
wondering in this case why xds cache does not even help mitigate the contention, we should make xds cache help

**Comment by howardjohn** [comment]
The service is scaling from 500->0, each time it does it invalidates the cache. So if the scaling is slow enough we will push EDS for the same service 500 times per connected proxy.

Cache cannot help unless we split the endpoints into chunks (LEDS)

**Comment by howardjohn** [comment]
> How many cores does the node istiod run on have? that will impact GOMAXPROCS.
I tried tuning this, minimal or no impact

**Comment by ra-grover** [comment]
> > How many cores does the node istiod run on have? that will impact GOMAXPROCS.
> > I tried tuning this, minimal or no impact

Sorry not sure which case are you mentioning ? Is it tuning the `PILOT_PUSH_THROTTLE` or your patch, for which you are asking the node cores. For tuning the `PILOT_PUSH_THROTTLE`, the node type used is `m5a.4xlarge`.

**Comment by howardjohn** [comment]
Sorry markdown was messed up. Was just commenting I tried tuning GOMAXPROCS on my own cluster and saw no benefits.

**Comment by howardjohn** [comment]
Some other ideas:

* Refactor queue to dedupe. no need to have events [`update endpoints to {a,b}`,`update endpoints to {a}`]; this can be collapsed down to a single event.
* Completely remove locks, use atomic map instead. This _almost_ works: https://github.com/howardjohn/istio/pull/new/xds/drop-locking-master-nolock. Might be able to make it fully work with some more efforts

**Comment by 4wdonny** [comment]
I'm seeing some symptoms that appear to be related to this. We have very bursty workloads and have experienced the 503's related to requests being sent to old/stale endpoints. We are using Istio `1.14.6` though.

**Comment by diranged** [comment]
@howardjohn This should still be open, right?

**Comment by howardjohn** [comment]
The original issue has essentially been resolved. So probably can be closed. if you are having issues on 1.14 -- upgrade. If you are having issues on modern versions with the fixes, probably best to open a new issue as this one is already huge, and its likely not exactly the same issue

**Comment by skizot722** [comment]
@howardjohn we ran into this same issue on a massive cluster with a very high amount of churn.  We were able to stabilize and improve performance by setting `PILOT_PUSH_THROTTLE` down to `15`.  It was originally unset, so I believe was using the default of `100`.

You linked in a [branch](https://github.com/howardjohn/istio/pull/new/xds/drop-locking-master-nolock) of yours above that completely removes the locks, which I believe is the culprit behind what we and @ra-grover encountered.  It looks like this change never made it into master?  Are there plans to remove the locking still?  You mentioned "The original issue has essentially been resolved" - how?  Do you mean resolved because of `PILOT_PUSH_THROTTLE` adjustment lower?

**Comment by howardjohn** [comment]
@skizot722 https://github.com/istio/istio/pull/45224 and https://github.com/istio/istio/pull/45230 drop the locking a lot, we don't need the complete removal IMO.

**Comment by skizot722** [comment]
Thanks @howardjohn - wasn't aware of those changes! Do you know what version they're expected to land in?

**Comment by ra-grover** [comment]
@skizot722 That fix is available in v1.19.0

**Comment by diranged** [comment]
Is it hard to back-port to the 1.18 codebase?

**Comment by altdelnow** [comment]
> One question before we get into it: have you used other Istio versions before and done similar things? if so, did you see the same issue? Trying to understand if its a regression or has always been any issue.

This issue of 503 is seen in all the version which we have used in past 3 years...1.12, 1.15 and now in 1.18.5. Because of this we are not using HPA and forced to go for fix number of PoDS based on max load which truly defeats the purpose of EKS. Any fix or parametric tweaking suggestion will be highly appreciated.

**Comment by howardjohn** [comment]
> > One question before we get into it: have you used other Istio versions before and done similar things? if so, did you see the same issue? Trying to understand if its a regression or has always been any issue.
> 
> This issue of 503 is seen in all the version which we have used in past 3 years...1.12, 1.15 and now in 1.18.5. Because of this we are not using HPA and forced to go for fix number of PoDS based on max load which truly defeats the purpose of EKS. Any fix or parametric tweaking suggestion will be highly appreciated.

All the fixes discussed in this thread are fixed **after** 1.18 FWIW.

**Comment by parshva-shah-dsw** [comment]
@howardjohn @ra-grover in our case the client application logs **"java.net.socketTimeoutException:readTimeout"** error when we have scale up or scale down of worker nodes. On further debugging we dont see any logs at istio gateway level. can you suggest on more debug steps to identify the root cause for the issue.

**Comment by ra-grover** [comment]
@parshva-shah-dsw We faced a similar problem, are you using the NLB in Instance mode or IP mode. We observed similar issues in Instance mode and they were mostly occurring during the node churn (scale up/down).

**Comment by parshva-shah-dsw** [comment]
@ra-grover we are using aws nlb ip mode and the target group is the EKS cluster node group. can we connect offline to understand more and discuss will be a great help

**Comment by parshva-shah-dsw** [comment]
@ra-grover @howardjohn do you require any more information?
